### ISSUE TITLE
[service] Add a patch_type build

### ISSUE THREAD

**Issue by SpaceIm** [conan-io/conan-center-index#14445]
Current allowed values `patch_type` miss the most common one: "build" fix.

conan refers to patch type we can't submit upstream
bugfix to change in source code
portability to fixes allowing to support more os, arch or compiler

But there is no patch type referring to a generic build fix which can be submitted upstream.

**Comment by SSE4** [comment]
what's the generic build patch? does it mean upstream doesn't build out of the box at all, for none of the configurations?

**Comment by SpaceIm** [comment]
For example a patch related to robustness, upstream build files is fragile, it builds when it assumes system installed dependencies, or it's not predictable.

**Comment by SSE4** [comment]
@prince-chrismc @uilianries WDYT? does it seem reasonable to allow to include some of these patch categories into conan-center, eventually?
I personally would just include that into `portability`, making this existing category definition wider, rendering also non-deterministic, non-predictable, fragile or somehow else insane build scripts (like using hard-coded paths to search for the dependencies and so on).
this new definition may include, besides already included new compiler/os/arch, also supporting different configurations (Debug vs Release vs RelWithDebInfo vs ...) or supporting other standard customizations, like `tools.build:cflags`/`tools.build:cxxflags`.
why? I think too many categories may easily confuse newcomers a lot, and open a debate or doubt on which categories particular patch falls into (as there are orthogonal dimensions, like can it be potentially submitted upstream vs does it modify build scripts or actual C++ source code).

**Comment by uilianries** [comment]
I agree with @SSE4, adding more one type will make it wider and harder to decide which type should be used, we usually have to guide newcomers with patch type and even for regular contributors. The `portability` covers most of cases, when adding new support when we need to add a new build support. I see you are asking about a patch type related to improvement/enhancement, something which should be submitted to the upstream. My vote would update portability patch type to cover that case, instead of creating a new one.

**Comment by SpaceIm** [comment]
> > Acually most of conan-center patches are not conan or portability types, just small improvements to make build robust.
> 
> Those will be rejected in the near future

What do you mean rejected? You don't want to have robust recipes in conan-center?

**Comment by prince-chrismc** [comment]
That was a terrible comment that needed more words. Sorry about that -- let me try again. I am not saying we have a clear definition but we want to _limit the patches_ (in the réduire sense of it). 

Looking at this from the highest point, I am trying to answer what are our "general principles" -- this topic is very nuanced so think of this as a starting pointing and then taking it case by case.

When it comes to 

> just small improvements

We should not accept patches that simply "make things nicer", we want to use the original build scripts (more generally the code) as much as possible and prefer making "conan corrections" in the recipe when feasible  (think changing install folders versus a move in the package method). Patches should be required to make it build, robustness is great if it's towards supporting more configuration or platforms. However if its towards making the build times faster thats not something we need per say (it's absolutely better and should be contributed back to the project). Again, case by case - there's lots of counter examples.

We have this idea that "source are immutable" but thats directly in opposition with "make it work everywhere" so we need to pick a point where both of those objectives are meet. I am trying gauge where that middle point is and put that in the docs.

When is something a patch? when is it a recipe detail?

**Comment by SpaceIm** [comment]
> We should not accept patches that simply "make things nicer", we want to use the original build scripts (more generally the code) as much as possible and prefer making "conan corrections" in the recipe when feasible (think changing install folders versus a move in the package method). Patches should be required to make it build, robustness is great if it's towards supporting more configuration or platforms. However if its towards making the build times faster thats not something we need per say (it's absolutely better and should be contributed back to the project). Again, case by case - there's lots of counter examples.
> 
> We have this idea that "source are immutable" but thats directly in opposition with "make it work everywhere" so we need to pick a point where both of those objectives are meet. I am trying gauge where that middle point is and put that in the docs.
> 
> When is something a patch? when is it a recipe detail?

When I mean small improvements, it's not about making build faster, but robust, and it's not a detail, it's usually not specific to conan, and has nothing to do with portability.
